source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs

```diff
@@ -0,0 +1,119 @@
+using Octopus.Server.Extensibility.Authentication.HostServices;
+using Octopus.Server.Extensibility.Authentication.Resources;
+
+namespace Octopus.Server.Extensibility.Authentication.DirectoryServices.IntegratedAuthentication
+{
+    class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
+    {
+        readonly ILog log;
+        readonly IAuthCookieCreator tokenIssuer;
+        readonly IAuthenticationConfigurationStore authenticationConfigurationStore;
+        readonly DirectoryServicesUserCreationFromPrincipal supportsAutoUserCreationFromPrincipals;
+        readonly IUserStore userStore;
+
+        public IntegratedAuthenticationHandler(ILog log, 
+            IAuthCookieCreator tokenIssuer,
+            IAuthenticationConfigurationStore authenticationConfigurationStore, 
+            DirectoryServicesUserCreationFromPrincipal supportsAutoUserCreationFromPrincipals,
+            IUserStore userStore)
+        {
+            this.log = log;
+            this.tokenIssuer = tokenIssuer;
+            this.authenticationConfigurationStore = authenticationConfigurationStore;
+            this.supportsAutoUserCreationFromPrincipals = supportsAutoUserCreationFromPrincipals;
+            this.userStore = userStore;
+        }
+
+        public void HandleRequest(HttpContext context)
+        {
+            var result = TryAuthenticateRequest(context);
+            
+            var principal = result.User;
+
+            // Decode the state object sent from the client (if there was one) so we can use those hints to build the most appropriate response
+            // If the state can't be interpreted, we will fall back to a safe-by-default behaviour, however:
+            //   1. Deep-links will not work because we don't know where the anonymous request originally wanted to go
+            //   2. Cookies may not have the Secure flag set properly when SSL Offloading is in play
+            LoginState state = null;
+            if (context.Request.Query.TryGetValue("state", out var stateString))
+            {
+                try
+                {
+                    state = JsonConvert.DeserializeObject<LoginState>(stateString.FirstOrDefault());
+                }
+                catch (Exception e)
+                {
+                    log.Warn(e, "Invalid login state object passed to the server when setting up the NTLM challenge. Falling back to the default behaviour.");
+                }
+            }
+
+            // Build the auth cookies to send back with the response
+            var authCookies = tokenIssuer.CreateAuthCookies(principal.IdentificationToken, SessionExpiry.TwentyDays, context.Request.IsHttps, state?.UsingSecureConnection);
+
+            // If the caller has provided a redirect after successful login, we need to check it is a local address - preventing Open Redirection attacks
+            if (!string.IsNullOrWhiteSpace(state?.RedirectAfterLoginTo))
```

Comment:
For another time: It sucks having this open-redirection attack code scattered through the different auth providers which use redirection. Upon reflection, it feels more like an Octopus Web Server concern common to quite a few auth schemes. I understand the reason why it's baked in here: The auth provider is taking responsibility to form up a Redirect HTTP response. Maybe a future PR could make it a shared responsibility? 🤔